examples/jwt/src/main.rs

```diff
@@ -0,0 +1,178 @@
+//! Example JWT authorization/authentication.
+//!
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p example-jwt
+//! ```
+
+use axum::{
+    body::{Bytes, Full},
+    extract::TypedHeader,
+    handler::{get, post},
+    http::{Response, StatusCode},
+    response::IntoResponse,
+    Json, Router,
+};
+use headers::{authorization::Bearer, Authorization};
+use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{convert::Infallible, fmt::Display, net::SocketAddr};
+
+// Quick instructions
+//
+// - get an authorization token:
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -d '{"client_id":"foo","client_secret":"bar"}' \
+//     http://localhost:3000/auth
+//
+// - visit the protected area using the token for authentication
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJiQGIuY29tIiwiY29tcGFueSI6IkFDTUUiLCJleHAiOjEwMDAwMDAwMDAwfQ.M3LAZmrzUkXDC1q5mSzFAs_kJrwuKz3jOoDmjJ0G4gM' \
+//     http://localhost:3000/protected
+//
+// - try to visit the protected area using an invalid token
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer blahblahblah' \
+//     http://localhost:3000/protected
+
+static JWT_SECRET: &[u8] = b"secret";
```

Comment:
Should probably have a warning about not doing this in production and a hint telling users what to actually do in prod.

Reply:
Or just do it better right away. I think this example should depend on `once_cell` either way to only construct the decoding key once, then it can also easily grab the secret from an env var in the initialization of the `DecodingKey`.

Reply:
A very simple way was used in the [`oauth` example](https://github.com/tokio-rs/axum/blob/main/examples/oauth/src/main.rs#L6): using environment variables. What do ou thing about to use this same approach in the JWT example?

Reply:
Yes, that's what I was saying :)

I guess it would work without putting the `DecodingKey` into a `once_cell::sync::Lazy` even, by putting it into a request extension (wrapped in `Arc<_>`) instead. Either way `DecodingKey` (and `EncodingKey` too) should only be constructed once, not once per request since constructing it is somewhat expensive AFAIK.

The jsonwebtoken docs for both types include:

> This key can be re-used so make sure you only initialize it once if you can for better performance

Reply:
@jplatte Tonight I'll try it, thanks a lot! :smiley:

Reply:
@jplatte, I'm trying to get the secret from an environment variable and initialize the `DecodingKey`/`EncodingKey` once, so suppose this struct:

```rust
#[derive(Debug)]
struct Keys<'a> {
    encoded: EncodingKey,
    decoded: DecodingKey<'a>,
}

impl<'a> Keys<'a> {
    fn new(secret: &'a [u8]) -> Self {
        Self {
            encoded: EncodingKey::from_secret(secret),
            decoded: DecodingKey::from_secret(secret),
        }
    }
}
```

what to you thing about this code using `lazy_static!`?:

```rust
lazy_static! {
    static ref JWT_SECRET: String = std::env::var("JWT_SECRET").expect("JWT_SECRET must be set");
}

lazy_static! {
    static ref KEYS: Keys<'static> = Keys::new(JWT_SECRET.as_bytes());
}
```

also, it can use `once_cell` instead:

```rust
static JWT_SECRET: Lazy<String> =
    Lazy::new(|| std::env::var("JWT_SECRET").expect("JWT_SECRET must be set"));

static KEYS: Lazy<Mutex<Keys>> = Lazy::new(|| Mutex::new(Keys::new(JWT_SECRET.as_bytes())));
```

Please let me know the right path to follow ...

Reply:
I think you don't need to propagate the lifetime of `DecodingKey` to the `Keys` struct, that lifetime parameter is going away with the next `jsonwebtoken` release anyways. You can just use `.into_static()` to get a `DecodingKey<'static>`, and I don't think the keys should be mutable (`Mutex` should not be necessary):

```rust
#[derive(Debug)]
struct Keys {
    encoded: EncodingKey,
    decoded: DecodingKey<'static>,
}

impl Keys {
    fn new(secret: &[u8]) -> Self {
        Self {
            encoded: EncodingKey::from_secret(secret),
            decoded: DecodingKey::from_secret(secret).into_static(),
        }
    }
}

static KEYS: Lazy<Keys> = Lazy::new(|| {
    let secret = std::env::var("JWT_SECRET").expect("JWT_SECRET must be set");
    Keys::new(secret.as_bytes())
});